### PR TITLE
[SPARK-48098][INFRA] Enable `NOLINT_ON_COMPILE` for all except `lint` job

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to enable `NOLINT_ON_COMPILE` for all except `lint` job.

### Why are the changes needed?

This will reduce the redundant CPU cycle and GitHub action usage.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass the CIs.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Hi, @gengliangwang . Could you review this PR when you have some time?

**Review by gengliangwang — APPROVED** [review]
LGTM. This can avoid duplicate lint checks

**Comment by dongjoon-hyun** [comment]
Thank you so much, @gengliangwang .
Merged to master.

**Comment by dongjoon-hyun** [comment]
Oh, sorry. I need to revise this PR.

File: project/SparkBuild.scala [hunk 1]
```diff
@@ -257,7 +257,7 @@ object SparkBuild extends PomBuild {
 
   lazy val sharedSettings = sparkGenjavadocSettings ++
                             compilerWarningSettings ++
-      (if (sys.env.contains("NOLINT_ON_COMPILE")) Nil else enableScalaStyle) ++ Seq(
+      (if (!sys.env.get("NOLINT_ON_COMPILE").contains("false")) enableScalaStyle else Nil) ++ Seq(
```
Comment (dongjoon-hyun) [inline project/SparkBuild.scala:260]: 1. no env variable => `enableScalaStyle`
2. `NOLINT_ON_COMPILE=1` => `enableScalaStyle`
3. `NOLINT_ON_COMPILE=true` => `enableScalaStyle`
4. `NOLINT_ON_COMPILE=false` => Nil

**Comment by dongjoon-hyun** [comment]
Could you review this PR once more, @gengliangwang ?

File: project/SparkBuild.scala (hunk 1 above)
Comment (gengliangwang) [reply project/SparkBuild.scala:260]: nit: Shall we rename to `LINT_ON_COMPILE` and only disables it when `NOLINT_ON_COMPILE=false`?  Seems easier to understand

Comment (dongjoon-hyun) [reply project/SparkBuild.scala:260]: Ya, I thought in that way, but it's required for backward compatibility.

Comment (dongjoon-hyun) [reply project/SparkBuild.scala:260]: The existing semantic has been used since SPARK-14790 for 8 years.

Comment (gengliangwang) [reply project/SparkBuild.scala:260]: ok, I don't have a strong opinion here.

**Review by gengliangwang — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you for review and approval.

**Pull request closed without merging**